### PR TITLE
log chat messages

### PR DESCRIPTION
**Pull request by Sokomine**
Many players including myself wish to be able to log chat messages. Reading a text file is way more convenient than scrolling up and down ingame. Plus it adds timestamps. Sometimes colors in chat may be hard to read. And sometimes conversations are more complex than exchanging "Hi"'s.

This PR adds the form of logging that I'm personally using since some years and that other players have asked for and included in their builds. It requires editing whenever the upstream file is changed. And some players who want chat logging lack the skill to patch the code themshelves.

The PR is not perfect. Improvements are welcome! In particular:
1. Logging to an extra file (perhaps a logfile for each day, perhaps even include server name) would be better than debug.txt
2. The "ACTION[Main]: [Chat]" marker is currently needed for finding/grepping such entries in debug.txt but makes the whole thing less readable. Extra files would be better.
3. There ought to be a settings option to turn logging on and off.
4. If possible it would be great to see which server the player connected to.

### CODE REVIEW

**Comment by Zughy** [comment]
I expect a controversial label and a discussion about an optional flag due to potential privacy concerns (and I guess being GDPR/CCPA ecc compliant). Unsubscribing

**Comment by oong819** [comment]
My brain have 2 stream of thinking:
<br>
## Favouring statements:
* Minetest are not a messaging software so player should not expecting chat privacy
* The server admin should know every chat messages to know and action when something bad happens (bullying, server rules violating stuff, etc.)
## Againts statements
* Some players might direct messaging sensitive information such as account password, secret base position... and outside of Minetest such as real life personal informations
* Minetest should respect privacy, even if its are just a game

<hr>
if this got merged, the player should know if chat logging are enabled

**Comment by SmallJoker** [comment]
I do not see any issue with the concept of logging.

1. Chatting implies that you agree that the server distributes the chat message to all clients.
2. Chat messages are already logged on server-side.
3. This PR is already possible to implement using a trivial CSM mod.

For reference, most IRC users log chat messages in two places: bouncer scrollback (temporary) and logs on client-side. Other chat apps may even preserve the history on the server and replay it to any new member as well. This is not an argument for or against this PR. Instead, refer to the numbered listing above.

**Comment by Zughy** [comment]
> I do expect server admins not to look at private chat messages unless by accident or when a player reports a problem with another player regarding chat messages

I don't agree: no one shall ever be trusted, server admins included. If I bully someone via private messages, that someone can still screen/record the conversation so that admins can analyse the situation and potentially take action. There should be at least an option for mods to avoid logging messages (e.g. for mods implementing private messages), maybe as a parameter in the API

**Comment by Sokomine** [comment]
[Zughy](https://github.com/Zughy) wrote:
> I don't agree: no one shall ever be trusted, server admins included.

That's a good approach in theory. I'm afraid it can't work with Minetest. If the server owner is at least halfway capable, he/she can edit the server code.

> If I bully someone via private messages, that someone can still screen/record the conversation so that admins can analyse the situation and potentially take action.

If things get hot between two sides, such screenshots may be faked. There's also the factor of context. Getting only part of a conversation with a screenshot may lead to a completely wrong impression.

> There should be at least an option for mods to avoid logging messages (e.g. for mods implementing private messages), maybe as a parameter in the API

Doesn't help. Your hypothetical untrustworthy server admin will change the code so that the flag gets ignored.

The only way I see around this is end-to-end clientside encryption. If you worry about that maybe you can write a CSM that does exactly that? And find a way that servers can't forbid that particular type of CSM? Or even add it to the client as such so that everyone can profit? That way the messages won't only be protected from the server admin (which I usually consider to be trustworthy, at least on the larger known servers) but from all those who log as much as they can inbetween and abuse their power.

But that's not what this mod is about. Even with end-to-end encrypted messages, the receiving player needs to be able to read the message somehow. I'm arguing for that player to be allowed to log those private messages he/she sends and receives. Plus normal chat.

I don't care if it gets logged on the server. I wouldn't mind at all if this logging option would be purely client-side. After all that's what it's for: Help people scroll back, support their memory. Like re-reading an old email or letter.

**Comment by rubenwardy** [comment]
So for a server owner it's actually much clearer - server owners should have privacy policies and disclose that they log messages. Then it's totally fine.

But this issue is about client-side logging. I'm not sure about this case. I believe it may be covered by "recording your own conversations" and so be perfectly fine privacy law wise. With log rotation, it also isn't as big of a deal

**Comment by tigercoding56** [comment]
> So for a server owner it's actually much clearer - server owners should have privacy policies and disclose that they log messages. Then it's totally fine.
> 
> But this issue is about client-side logging. I'm not sure about this case. I believe it may be covered by "recording your own conversations" and so be perfectly fine privacy law wise. With log rotation, it also isn't as big of a deal

i think maybe only log private messages send to client that is logging (player could already screenrecord text if they wanted to)

**Comment by Zughy** [comment]
(RE @Sokomine and server admins: true, my approach was just to get rid of non tech-savvy people. I mean, there's no need to change the client, I can just write a mod to log everything if I want to (I think?))

**Comment by SwissalpS** [comment]
> (RE @Sokomine and server admins: true, my approach was just to get rid of non tech-savvy people. I mean, there's no need to change the client, I can just write a mod to log everything if I want to (I think?))

Server-side yes, a lot can be logged with mods. Client-side not without building modified client. OP is looking to add client-side logging.

> I still think that there should be a way to avoid logging messages, e.g. private messages . It's not really private if someone else outside the conversation can read it.

I agree that clients shouldn't be sent PMs not addressed to them. (If that currently is the case, we have another issue at hand with greater importance)

**Comment by hlqkj** [comment]
It may be off-topic here, but I'd like this feature to be scriptable (read, CSM or whatever you want to call them) and/or outputted to a separate file (w/ appropriate settings, of course).

TL;DR: privacy concerns here are actual, but they affect minetest in a way more generalized way, and in the context of this issue/PR they are just being brought up as a baseless justification to discourage the proposed feature.

<details>
<summary>My reasoning below.</summary>
 
**About privacy concerns.**
I have to support everything that has been said about privacy concerns: however, in my view those should be a matter for **servers** and their admins' policies, which should be clear enough about what (if) gets logged, where it is stored, and for how long, etc. in accordance with relevant and applicable laws (read, GDPR).

A player using a server is inherently accepting the server policies, them be explicitly stated or not, incl. personal data retention (and this isn't just the chat BTW, but technically whatever action a user do on a minetest server - the world itself is a massive log of the user activity after all). By the most strict law interpretation (disclaimer: I'm not a lawyer...) this process should in theory be explicit and require the user consent, while also providing them with some mandatory statements/information, like e.g. how their data is handled, by who, where it is stored, for how long, for what reasons, who is responsible of their data handling, what are their rights under applicable laws, and how they may exercise those rights, etc. In other words, stuff like that.

(Re-disclaimer: I'm not a lawyer...) Given the fact that companies like Micro$oft, Meta, Alphabet, Github, ... and services like IRC, Discord, ... (you got it) that act as either relay and/or retentive services do exist, I suppose there is a way for those "Terms and Conditions" to account for public chat in minetest servers as well! I mean, I could take a copy of this conversation; screenshot it; put it on some piece of something with a laser engraver, or print it with a dot matrix printer, and that's it!? I doubt any of you could sue Microsoft for this? (Please, do correct me if I am wrong here.)

The same applies to PMs. Once a user sends them, **unless** minetest explicitly advertised a feature by which the message is **not** screenshot-able, copyable, can be seen only once and for X seconds then it will self-destroy, then that user do accepts that their direct message _may_ be stored by the legit(*) recipient!

(*) _I've always assumed, and I **really** do hope I wasn't wrong for all this time, that `minetest.chat_send_player` is not so dumb to send the chat to all clients and let them clients filter messages out - if so, we have a greater problem somewhere else._

Speaking about PMs, this is the very only reason I'd ever wanted client-side chat logging in minetest. While most servers do publish their chats in some way, so I can get there and find some information I need at a later time, PMs of course are not shown there and hence, once I close my client, the chat scrollback is lost forever, along with the PMs that I received. CSM can't help, and screenshots ... oh well.

So, the very only concern related to privacy I may see in minetest, is that almost no server makes this (_required-by-law?? Remember, I'm not a lawyer!_) process explicit, e.g. by showing a "Terms and Conditions" form that must be accepted by new players at the registration stage.

But for how things are here and now, privacy really doesn't matter for this PR.

~~_My personal opinion as an EU citizen: this is all c**p. Or, the way this has been exacerbated, is... And I apologize to the rest of the world for all the cookie banners as well! :@_~~

**About it being scriptable, and/or sent to separate file.**

1. Once we grant this is legit, the average player would be annoyed if they have to learn how to use `grep` to get their chat out of debug.txt - and I would be too.
2. Not using grep makes the debug output useless when it comes to chat, as it legit include other stuff which may be a lot of stuff when for some reason the verbosity is increased.
3. It makes life easier when it comes to delete/rotate debug.txt and chat log if they are separate.
4. Someone may like to be able to print their chat logs in a fancier way than the ugly debug output, without having to learn `sed` as well.
5. I'd also like to log my current position as well with every chat message, as I do on my servers chat logs.
6. Ever happened to play different servers with the same client, maybe at the same time? Chats would be mixed up.

I know this will now summon the worse havoc, but why don't just finally let CSMs write text files (ofc, with safety in mind - e.g. only ASCII printable characters, etc.) and have the client-side chat logging handled by `builtin`?

Apologies and thanks to those who took the time to read till here :)
</details>

**Comment by Sokomine** [comment]
@rubenwardy, @SwissalpS: Thanks for clarifying the topic! I ought to have thought of that from the beginning.

[Zughy](https://github.com/Zughy) wrote:
> (RE @Sokomine and server admins: true, my approach was just to get rid of non tech-savvy people. I mean, there's no need to change the client, I can just write a mod to log everything if I want to (I think?))

I hope that server admins are sufficiently tech-savvy. Else they can't really manage to run a server well. But this is kind of explicitly for less tech-savvy people. They may want a log as well - and they don't know how to modify their client to do it.

> I still think that there should be a way to avoid logging messages, e.g. private messages (see [this mod](https://content.minetest.net/packages/SonoMichele/betterpm/)). It's not really private if someone else outside the conversation can read it.

No way around that apart from encryption. The server owner can always modify the server and/or mod, and all those governmental organizations that don't give a damn about privacy or human beeings log whatever they can anyway.

[hlqkj](https://github.com/hlqkj) wrote:

> It may be off-topic here, but I'd like this feature to be scriptable (read, CSM or whatever you want to call them) > and/or outputted to a separate file (w/ appropriate settings, of course).

No, that's not offtopic at all. That's just an additional (sensible) feature request. It's far more on-topic than most of the rest of the discussion here.

> The same applies to PMs. Once a user sends them, unless minetest explicitly advertised a feature by which the message is not screenshot-able, copyable, can be seen only once and for X seconds then it will self-destroy, then that user do accepts that their direct message may be stored by the legit(*) recipient!

Exactly. Only very few chat systems advertise such self-destroying features. Many of us use IRC for example. It's just a config option in most clients if you want to log private conversations or not.

> (*) I've always assumed, and I really do hope I wasn't wrong for all this time, that minetest.chat_send_player is > not so dumb to send the chat to all clients and let them clients filter messages out - if so, we have a greater  problem somewhere else.

There's no indication that there's something wrong there. I've run that client-side logging for a while now and didn't receive any private messages that were not for me (apart from those where the sender mistyped the name of the receipant or confused receipants).

> (1.-3.)

Yes. A seperate log file would be far better. Creating such a one at that point exceeded my abilities. I was glad to have found the right location in the code to get the information - in src/network/clientpackethandler.cpp, right before it gets sent to CSM and either consumed by a CSM or put in the chat queue for display on the client. As long as CSM are not forbidden (they are forbidden on most servers for *other* reasons than logging your chat), a CSM could indeed do the work here. Provided they can write files.

> 4. Someone may like to be able to print their chat logs in a fancier way than the ugly debug output, without having to learn sed as well.

Definitely. I'm wondering if some sort of simplified HTML might be helpful. The chat lines at that point contain some additional information like color and sometimes source of the message (player died, used anvil, ...).

> 5. I'd also like to log my current position as well with every chat message, as I do on my servers chat logs.

Sure, why not. I just have no idea how to get that information at that point. For me personally it'd be less intresting as I'm often just standing somewhere while chatting. But could indeed help if someone wants to find an intresting location again.

> 6. Ever happened to play different servers with the same client, maybe at the same time? Chats would be mixed up.

Good point! Logging each server to a diffrent logfile might be the best solution.

There are also those commands that the player typed and sent to the server. If the server doesn't echo them back to the client they're not caught by my simple one-line-PR.

> 7. Using LUA makes it possible to do some filtering as well, by e.g. only logging messages from specific names.

Perhaps that's really done best in a client-side script then?

> 8. What about server messages, sometimes they matter, too.

Some specific server messages may even be the reason for some players to be intrested in chat logging. On the Your Land server, there are events where an...entity..."speaks" in chat in very dark color. Those are highly entertaining server events - but it's hard to read. Plus those players who weren't there want to know later on what the entity said (RPG, progress of a story). I'm..pretty...certain that the resident evil overlord of a server (an RPG element) doesn't mind beeing logged :)

Technical help to get this PR usable would be really appreciated. There needs to be a way for normal players to switch it on and off, and a seperate logfile (per server) would be really good.

**Comment by hlqkj** [comment]
@Sokomine:

> There's no indication that there's something wrong there. I've run that client-side logging for a while now and didn't receive any private messages that were not for me (apart from those where the sender mistyped the name of the receipant or confused receipants).

I got worried just because of people here saying they're concerned for "third-party" PMs ending up in the client chat logs your PR implements, whereas in my understanding that would not have been possible at all, being that the client does not even receive such PMs: thanks for confirming this!

> Definitely. I'm wondering if some sort of simplified HTML might be helpful. The chat lines at that point contain some additional information like color and sometimes source of the message (player died, used anvil, ...).

That'd even be too much, and perhaps too complex as well.

<details>
<summary>About the use of HTML...</summary>
Consider that an HTML file would have to have a fixed header part, then messages appended and have a footer closing it. This means that you can't just grow the file by appending new messages to it.

Either you create a new log file per server, per session; initialize it at the beginning, growing it by adding line, and finalizing it at the session end. However, the resulting file would _not_ be a valid HTML before the session ends and even end up being corrupted if the session ends abnormally.

Or, you would have to maintain valid HTML file by updating it with a seek-and-append approach, which is ~~inefficient~~ boring.
</details>

I was thinking about something simpler and more elegant like this (taken from my server' logs, redacted, and inspired by someone else's work, someone I don't know if they wants to be mentioned here).

<details>
<summary>Log format example</summary>

```
[2024-01-19 22:07:57 @ (983,11,837)]              <A****> Np
[2024-01-19 22:08:12 @ (983,13,851)]              <A****> Gotta hurry befor I starve
[2024-01-19 22:08:19 @ (990,11,848)]              <-p********> ok
[2024-01-19 22:09:00 @ (971,1,942)]               <A****> Help
...
[2024-01-23 08:43:05]                             # Server: Normal shutdown. Everybody off!
[2024-01-23 09:02:12]                             # Server: Startup complete.
[2024-01-23 09:02:17]                             # Server: The border is open. New players are allowed to join.
[2024-01-23 12:06:07 @ (911,13,599)]              *** <-p********> joined the game.
```

</details>

Of course, if the decision is made to let this feature scriptable, more complex use cases could be accomplished. For example, if this ends up being implemented in `builtin`, an API could be exposed to format the messages before they get logged, and maybe even more API to initialize and finalize a log file (making your life easy if you _really_ need HTML). Or, just open a bit the CSM environment and let people write their logging scripts to their needs! (<-- even easier HTML :)

> Sure, why not. I just have no idea how to get that information _[player's position]_ at that point. For me personally it'd be less intresting as I'm often just standing somewhere while chatting. But could indeed help if someone wants to find an intresting location again.

That'd again be trivial if done on the LUA side.

> There are also those commands that the player typed and sent to the server. If the server doesn't echo them back to the client they're not caught by my simple one-line-PR.

I didn't consider this, much needed indeed!

> Some specific server messages may even be the reason for some players to be intrested in chat logging. On the Your Land server, there are events where an...entity..."speaks" in chat in very dark color. Those are highly entertaining server events - but it's hard to read. Plus those players who weren't there want to know later on what the entity said (RPG, progress of a story). I'm..pretty...certain that the resident evil overlord of a server (an RPG element) doesn't mind beeing logged :)

That's indeed what I mean, or messages sent by the server to the player and not part of the public chat (logs). E.g. a specific server I play has a feature that lets the player hide their death message, which there contains the coordinates to their coffin/bonebox. This message is then sent privately via a server DM and, if lost, the way to the goodies may also be lost.

> Technical help to get this PR usable would be really appreciated. There needs to be a way for normal players to switch it on and off, and a separate logfile (per server) would be really good.

I already tried to implement this on client-side, but given the _silly_ way the CSM environment limits filesystem access, all my attempts ended up hitting a wall. I am not sure about what exactly is available in `builtin`'s environment when running as a client, however, if a very basic access is (or will be) available I could try sketching something.

Another point to ponder: if done by CSM alone, some servers (like Your Land) completely block CSM loading. This could be a contradiction, in a situation where a feature is advertised as part of the client but then it gets silently turned off by specific servers. ~~_And here we could again discuss about the whole point of CSM restriction flags, and limiting the CSM environment to the point where I can't write a text file in a specified folder, when there is a nice "dragon firing around"... :-)_~~

**Comment by appgurueu** [comment]
As an issue / feature, this has my support. It is not our problem whether this complies with local data protection regulations, so long as Minetest doesn't make compliance harder per se.

I agree with the four proposed improvements. I think *at least* the setting *needs* to be implemented, everything else is important to have but not strictly necessary.

**Comment by Sokomine** [comment]
[hlqkj](https://github.com/hlqkj) wrote:
> Consider that an HTML file would have to have a fixed header part, then messages appended and have a footer closing it. This means that you can't just grow the file by appending new messages to it. 

In theory such a HTML file without closing tags at the end would be invalid HTML. In practice, most browsers don't insist that strongly on correct HTML and can often deal with such files without problems. Admittedly it's dirty. But might help a lot of users who might be confused with text with color tags.

> E.g. a specific server I play has a feature that lets the player hide their death message, which contains the coordinates to their coffin/bonebox. This message is then sent privately via a server DM and, if lost, the way to the goodies may also be lost.

Yes, that's another important use case for many.

[appgurueu](https://github.com/appgurueu) wrote:
> I agree with the four proposed improvements. I think at least the setting needs to be implemented, everything else is important to have but not strictly necessary.

That's right. The settings parameter is essential. The rest would be highly welcome (i.e. also logging what the player sent to the server and which may not always be echoed back).

> If you need help with the implementation, feel free to ask.

I do need help with that. I was able to locate the right place to plug in and get it working for me, but doing it properly for all in a way that we can all be comftable with if it gets merged is a bit beyound my capabilities regarding C, C++ and in particular the MT source code at the moment. I could probably figure out handling a setting as such, but then there's the issue of not logging to debug.txt. I have no idea about how to do at the moment.

> Somewhat related: https://github.com/minetest/minetest/pull/14290. If we have chat logging, we can load the appropriate chat logs when a user joins a server.

Perhaps. But that issue is mostly about having more lines to scroll back through without having to resort to the log yet. Adding a config option for that is of course technicly very similar to a config option for enabling logging.

Passing the data on to builtin as [hlqkj](https://github.com/hlqkj) suggested might be a good solution.

**Comment by hlqkj** [comment]
I think issues should be kept separated: #14290 needs an internal "raw" storage for the chat, which includes all the (color) escape sequences and anything else that may be needed by a good scrollback implementation (eg. typed lines history?! I don't know how that should work). This can be handled by the engine itself and left as simple and clear as possible.

As per user logs, those could be implemented in LUA in builtin and let it be scriptable via CSM (provided there is a way to do that). We can decide for a default export format, either color-stripped plain text or HTML or anything else, while also providing exotic users like me the possibility to do more custom stuff.

**Comment by oong819** [comment]
Now i'm very want this to be merged. Currently i have to screenshot every single chat...

File: src/network/clientpackethandler.cpp [hunk 1]
```diff
@@ -435,6 +435,9 @@ void Client::handleCommand_ChatMessage(NetworkPacket *pkt)
 
 	chatMessage->type = (ChatMessageType) message_type;
 
+	// log the chat message
+	actionstream << "[Chat] " << wide_to_utf8(chatMessage->message) << std::endl;
```
Comment (hlqkj) [inline src/network/clientpackethandler.cpp:439]: The tag string should be `"CHAT: "` for consistency, see: https://github.com/minetest/minetest/blob/fbec168e91026b70fd989e7f46bea7140328076e/src/server.cpp#L3115.

**Comment by sfan5** [comment]
So whats holding this up?

**Comment by nerzhul** [comment]
Wide question, should server set a flag to client to permit chat logging, as people talk about privacy ?
I'd like also to see an option client side to enable or disable it. In my case, i don't care about using disk space for chat messages

File: src/network/clientpackethandler.cpp (hunk 1 above)
Comment (sfan5) [inline src/network/clientpackethandler.cpp]:
```suggestion
	actionstream << "CHAT: " << wide_to_utf8(unescape_enriched(chatMessage->message)) << std::endl;
```

**Review by sfan5 — APPROVED** [review]
